Faster defaults

Shnatsel opened the pull request:
* Default to compiling in release mode with overflow checks and debug assertions enabled
    * Speeds up fuzzing ~10x
* ~~Add command line option for controlling overflow checks~~
* Add command line option for forcing debug mode without optimizations

Resolves #230

fitzgen commented on src/options.rs:
```diff
@@ -57,18 +57,26 @@ impl FromStr for Sanitizer {
 #[derive(Clone, Debug, StructOpt)]
 pub struct BuildOptions {
-    #[structopt(short = "O", long = "release")]
+    #[structopt(short = "D", long = "debug", conflicts_with = "release")]
+    /// Build artifacts in debug mode, without optimizations
+    pub debug: bool,
```
Let's name this `--dev` to match the cargo profile name, and to avoid confusion like "how is this different from `--debug-assertions`?"

Shnatsel replied:
Done

fitzgen commented on src/options.rs:
```diff
@@ -57,18 +57,26 @@ impl FromStr for Sanitizer {
     pub debug_assertions: bool,
 
+    #[structopt(short = "o", long = "overflow-checks")]
+    /// Build artifacts with overflow checks enabled (default if not -O)
```
Regarding "default if not -O": right now, passing `-Oa` will enable overflow checks, is that not the case anymore after this PR? Many projects use `-Oa` and I'd be wary of making it lose the overflow checks. Do `debug-assertions` imply `overflow-checks` or are they disconnected now?

Shnatsel replied:
> right now, passing -Oa will enable overflow checks, is that not the case anymore after this PR?

That is still the case. The way these flags work is unchanged. Only the default state is altered, which is now equivalent to `-Ooa`

> Do debug-assertions imply overflow-checks or are they disconnected now?

I believe debug-assertions never implied overflow-checks, and enabling overflow-checks in release mode was not possible prior to this PR.

Shnatsel replied:
Oh wait, nevermind, I was confused. I'll have to test whether `-a` actually enables overflow checks or not.

Shnatsel replied:
Okay, I've tested it and turns out `--debug-assertions` already enables overflow checks, so there's no real need to have a separate flag for it. I'll drop the extra flag.

fitzgen commented on src/project.rs:
```diff
@@ -507,7 +507,7 @@ impl FuzzProject {
 
             eprintln!(
-                "Reproduce with:\n\n\tcargo fuzz run {target} {artifact}\n",
+                "Reproduce with:\n\n\tcargo fuzz run --debug {target} {artifact}\n",
```
Let's leave these without any flags for simplicity, and handle adding flags for them in #228

Shnatsel replied:
Done

fitzgen commented:
Tests are failing because they [look for bins at the `target/debug` path](https://github.com/rust-fuzz/cargo-fuzz/blob/87c36a1deee8ca9b550fe034623a7445d7e8c770/tests/tests/main.rs#L622) in various places, so they need to update to `target/release`.

Also, please add a test for the `--dev` (ne `--debug`) flag.

Shnatsel commented:
I've fixed everything as requested. Will add a test for `--dev` shortly.

Do you have any ideas on how to clearly document that the default is `--release --debug-assertions --overflow-checks`? If I invoke `--help` it's not clear to me what the build mode will be if no parameters are passed.

Also, there are two changes that might be unexpected for end users:

 - This changes the coverage numbers. Now that a great deal of branches can be optimized out, the reported coverage is much lower (~2000 instead of ~7500 edges for `png` crate).
 - This changes the default location of the binary. If something was invoking the generated binary directly instead of via `cargo fuzz run`, that workflow will break.

Should we document this somewhere?

Shnatsel commented:
Should be good to go now. Please take another look.

fitzgen commented:
> Also, there are two changes that might be unexpected for end users:

I'm personally not too concerned with either of these, but we can bump to 0.8.x to be safe. @Manishearth any opinion here?

> Do you have any ideas on how to clearly document that the default is `--release --debug-assertions --overflow-checks`? If I invoke `--help` it's not clear to me what the build mode will be if no parameters are passed.

Perhaps in the before-help blurb for `cargo fuzz run`? https://github.com/rust-fuzz/cargo-fuzz/blob/87c36a1deee8ca9b550fe034623a7445d7e8c770/src/main.rs#L36-L43

And I think we would also want a custom template for `cargo fuzz build` like how `cargo fuzz run` has: https://github.com/rust-fuzz/cargo-fuzz/blob/87c36a1deee8ca9b550fe034623a7445d7e8c770/src/main.rs#L107

This way we could include the same blurb for both subcommands's help messages.

nagisa commented:
> I'm personally not too concerned with either of these, but we can bump to 0.8.x to be safe.

There's not really a good reason to _not_ bump a "breaking" version number for binaries once there are larger changes to the interface. I vote bump.

Shnatsel commented:
I've added a blurb about default build flags to the help text of both `cargo fuzz build` and `cargo fuzz run`. I've also added a blurb about sanitizers to `cargo fuzz build` after-help text, mostly so that I wouldn't have to create yet another help template.

Manishearth commented:
@fitzgen I don't consider these to require a new major version

frewsxcv commented on src/main.rs:
```diff
@@ -68,6 +72,17 @@ include:
 
+const BUILD_BEFORE_HELP: &'static str = "\
+By default fuzz targets are built with optimizations equivalent to
+`cargo build --release`, but with debug assertions and overflow checks enabled.
```
> but with debug assertions and overflow checks enabled

which code in this pr is responsible for enabling debug assertions with release mode by default?
